### DISCUSSION TITLE
[RFC] Addons

### DISCUSSION THREAD

**Discussion by pontusab** [resend/react-email#1044] [Ideas]
### Goals

1. Extend react-email with plugins (i18n helpers)
2. Extend react-email preview with controllers for example i18n selector to choose language

### Background

Im currently implementing i18n within each email, I would like a config where I can define languages and translations where can toggle between them within the preview tool.

### Config

`Welcome.config.ts|tsx`

```js
import { withI18n, withConfig } from '@react-email/addons';
import { Welcome } from './Welcome';


const defaultProps = {
    firstName: 'Pontus'
}

export default withConfig(withI18n(Welcome), defaultProps)

```

## Translations
In `emails/welcome` add `translations.ts`

```js
export default {
    en: {
        welcome: 'Welcome {firstName}!,
        button: 'Click me',
    },
    sv: {
        welcome: 'Välkommen {firstName}!',
        button: 'Klicka på mig'
    }
}
```

### Template

`Welcome.tsx`

```jsx
import { Html, Button, Heading } from "@react-email/components";
import { I18nProps } from '@react-email/addons';

type Props & I18nProps = {
    firstName: string
}

const Email = ({ t, locale, firstName }) => {
  return (
    <Html lang={locale} dir="ltr">
    <Heading as="h2">{t('welcome', { firstName })}</Heading>

      <Button href="https://example.com" style={{ color: "#61dafb" }}>
        {t('button)}
      </Button>
    </Html>
  );
};

```

**Comment by zenorocha** [comment]
That's super promising. Is there a good reference you could point to for this?

**Comment by pontusab** [comment]
We would then need to require the `Welcome.config.ts` file instead of the template to get the hoc in runtime. One other way would be to introduce a global config like `email.config.ts` and rethink how the template is included

**Reply by pontusab** [reply]
Storybook is a good example: https://storybook.js.org/docs/react/addons/writing-addons
